src/librustc_mir/transform/uniform_array_move_out.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This pass converts move out from array by Subslice and
+// ConstIndex{.., from_end: true} to ConstIndex move out(s) from begin
+// of array. It allows detect error by mir borrowck and elaborate
+// drops for array without additional work.
```

Comment:
I think that, to document this sort of thing, it is best if we can start with a comment that includes some kind of "before/after" -- e.g., show the MIR that we are starting with, and the MIR that we are ending with. Then, in the actual code, we can leave comments that reference this running example.